### ISSUE TITLE
PulseEffects + OBS desktop audio captures microphone audio instead

### ISSUE THREAD

**Issue by Art1f4ct** [wwmm/easyeffects#807]
Whenever PulseEffects is configured to process microphone audio captured by OBS, the desktop audio in OBS audio mixer becomes a duplicate of microphone. In PulseAudio Volume Control I can see that both OBS recording devices use "Monitor of PulseEffects(mic)". Changing the audio device in OBS settings has no apparent effect. However, using PulseAudio Volume Control I can change OBS Desktop Audio to record stream from "Monitor of PulseEffects(apps)" which fixes the issue until OBS is closed.

Since OBS desktop audio is detected as a recording device I suspect that PulseEffects assumes OBS is trying to capture microphone audio and forcibly redirects both OBS devices to PulseEffects mic. When I try to blocklist OBS desktop audio PulseEffects also blocklists OBS mic which means OBS mic audio doesn't get processed.

**Comment by wwmm** [comment]
Is the option `Process All inputs` enabled? If yes try to disable it. After that PulseEffects won't force applications to record from the `PulseEffects(mic)` monitor. You still might need to do the manual selection at least once because Pulseaudio remembers the last device used and try to automatically move the program to it again.

**Comment by Art1f4ct** [comment]
Yes, process all inputs is enabled (for convenience). Turning it off fixes the issue. However this isn't practical for my use case where it takes just as much or more effort to enable mic processing per application as it does to override OBS desktop audio. I realize I should have targeted the blocklist as my issue, which seems to be limited to a per-application basis. Expanding it to a source-based system seems like the ideal fix to my exact issue, although given how niche my use case is I'll just make a script that sets OBS desktop audio automatically and call it a day.

**Comment by ShayBox** [comment]
Would you mind sharing this script if you made it

**Comment by ShayBox** [comment]
This issue needs to be re-opened, with the new pipewire backend you can't change it with pavucontrol like before, there's no way to use pulseeffects with input processing enabled and obs.

**Comment by wwmm** [comment]
> with the new pipewire backend you can't change it with pavucontrol like before

This sounds like a bug in PipeWire's compatibility layer for Pulseaudio applications.

There is a related discussion in https://github.com/wwmm/pulseeffects/issues/822. The problem with OBS is that we will have to write code that is useful only to it. An idea I do not like...

**Comment by ShayBox** [comment]
I don't doubt that's a problem with pipewire not allowing me to manually switch it like I could with pulseaudio, but that's not the problem, that's the problem preventing me from using the workaround I was using, that was a way to get it working until pulseeffects switched it back 5 minutes later, and #822 would work if you just wanted obs to record your raw microphone, but as the intent is to use pulseeffects with obs, that would also not be a solution, obs with pulseeffects is probably the most common combination of programs any linux streamer is going to use, almost any studio microphone is going to need pulseeffects to sound acceptable

**Comment by wwmm** [comment]
In your case would not be enough to add OBS to the blacklist in the microphone section? It is not clear in my mind how you audio routing is configured

**Comment by wwmm** [comment]
Or is OBS recording from 2 devices?

**Comment by ShayBox** [comment]
Replicate: Install pulseeffects (pulseaudio or pipewire), enable input processing, install obs.

Microphone is correctly in the microphone source on obs, but the desktop and microphone are mixed together on the desktop source, there was a workaround using pavu with pulseaudio but now that doesn't even work, every single system and install i've tried has had this problem, I just flat out can't record with pulseeffects open unless I want don't want to record desktop audio

**BLACKLIST DOES NOT WORK**
**NOTHING** fixes this other than removing pulseeffects

**Comment by wwmm** [comment]
@ShayBox show some Pavucontrol and PulseEffects window images. As I have never used OBS it is not clear at all why the per application switches in PulseEffects window are not enough to select which OBS stream should or not have effects set. It should just be a matter of not enabling `Process All Inputs` or `Process All Outputs` in PulseEffects settings menu and using the switch button for each stream.

**Comment by ShayBox** [comment]
Not using input processing does fix it, that kinda defeats the point though, I need input processing for my microphone to be audible at all, and output processing doesn't fix it https://youtu.be/DTd7NjKvJAs

**Comment by Art1f4ct** [comment]
@wwmm OBS has two separate audio captures, one for microphone and one for desktop audio. PulseEffects forces both to capture PulseEffects mic if OBS is deemed by PulseEffects to be processed (this is when input processing is enabled and OBS is not on the blocklist). This means desktop audio is captured incorrectly.

Disabling input processing does fix it but that means you have to manually switch it on for every application using your mic. Ew! 

This most likely isn't limited to OBS. OBS is the easiest example we have to test, since applications such as Discord don't do desktop audio capture on Linux yet, but I suspect that PulseEffects input processing will interfere with /any/ application that attempts to grab desktop audio and microphone audio at the same time.

My solution is to allow the blocklist to target a specific source/stream? (idk the right terminology) name, rather than the application as a whole. This way it isn't a workaround for OBS specifically, but rather, it allows the user to configure more specific rules for what streams should/shouldn't be automatically processed. Desktop capture would still be broken until the user configures the blocklist though.

**Comment by wwmm** [comment]
>My solution is to allow the blocklist to target a specific source/stream? (idk the right terminology) name, rather than the application as a whole. 

That will face all the problems discussed at #822. There are many stream properties that Pulseaudio and PipeWire allows the application to set. Usually audio applications do not make a good use of any of them. OBS included. Instead of setting an unique id for each kind of stream it creates it just set the `media name` property to something different. But other applications use the media name property for totally different purposes. There is no way a generic filter can be made based on this without the risk of breaking other applications. And to make things worse the media name is something that can be translated... So the filter has to take this kind of thing into account...

In the current situation the only property that is really reliable is the application name. That is why the blocklist uses only it. In order to forbid OBS from desktop stream from being moved to our source I will have to resort to the worse possible solution. Hard coding rules specifically targeted to OBS and hope they do not change this stream media name. And that translations do not get in the way =/

**Comment by wwmm** [comment]
>application that attempts to grab desktop audio and microphone audio at the same time.

Probably. The only property applications developers are forced to set is the application name. So there is no reliable way to know the purpose of its streams and react accordingly. For each app deciding to do what OBS does a different solution will have to be made because I really doubt they will tag their streams with the same type of property... If they tag them at all =/

**Comment by Art1f4ct** [comment]
Is it possible to observe what audio pipeline these streams attempt to use before being moved by PulseEffects? If it is, maybe PulseEffects could distinguish between streams using the monitor of microphone versus monitor of desktop audio.

**Comment by wwmm** [comment]
>monitor of desktop audio

By this do you mean the sound card device monitor? Which device OBS desktop stream records when PulseEffects is not running?

Although it is not a straightforward process it is possible to know if a stream is recording a hardware source or a virtual monitor source. It can be tricky but it is not impossible. The problem is that when the system does not have a mic plugged PipeWire moves recording streams to any virtual source available. So there is that to consider too.

**Comment by wwmm** [comment]
There are also the cases where our virtual source could have been set as default source. In this case there is a high chance that PipeWire will link the stream to our virtual source as soon as it starts. So a logic based on hardware vs virtual would break. In this case it has to be through a blocklist.

**Comment by wwmm** [comment]
I am not sure when because the gtk4 port is taking much more time than I had anticipated but I think I will have to hard code rules for OBS. It does not look like there are easier alternatives that do not break other apps.

**Comment by Art1f4ct** [comment]
> By this do you mean the sound card device monitor? 

I'm not sure; I don't know much about the inner workings of Linux audio. All I really understand is that in PulseAudio Volume Control in the recording tab I can see the two OBS streams and I can see what device each stream is getting audio from. I know that the microphone is a hardware device but idk if OBS somehow creates the monitor of desktop audio device or if it's a part of PulseAudio/PipeWire or if it's a hardware device.

> There are also the cases where our virtual source could have been set as default source.

But this isn't how PulseEffects process all inputs option works right? I assume you're describing when people use a virtual device for other reasons, in which case I see why the differentiation of virtual/hardware devices wouldn't work.

Either way, even if it is somehow possible, it sounds like a lot of stuff could break if too much is changed. Maybe hard coding rules for OBS is the best option.

**Comment by wwmm** [comment]
>and I can see what device each stream is getting audio from

That is it. Which device does Pavucontrol shows as the source of OBS desktop stream?

>But this isn't how PulseEffects process all inputs option works right?

It isn't. The process all inputs options forces any recording stream to use our virtual device as source. What I was talking about before is that when no mic is plugged the stream will be moved to any virtual device available. It does not have to be one created by the user. Once a mic is plugged it can happen that the stream stays in the virtual device. So a solution based on whether the previous connection was in a virtual or real device would fail.

>Either way, even if it is somehow possible, it sounds like a lot of stuff could break if too much is changed. Maybe hard coding rules for OBS is the best option.

Yes. It is annoying but it is probably the less problematic solution.

**Comment by Art1f4ct** [comment]
> Which device does Pavucontrol shows as the source of OBS desktop stream?

For me it shows "Monitor of Built-in Audio Analog Stereo". "Bulit-in Audio Analog Stereo" is the name of my motherboard's internal audio, however, and if I were using HDMI as the default audio output before starting OBS it would use "Monitor of [really long HDMI device name]" instead.

Also I just tested on an Ubuntu live usb and I can confirm that the "Monitor of [Device]" devices are not made by OBS, they immediately exist on the system.



Below is a screenshot showing PulseAudio Volume Control, after opening OBS, without PulseEffects running.

![Screenshot from 2021-04-12 01-39-33](https://user-images.githubusercontent.com/73558637/114331002-95461400-9b32-11eb-9441-9a2019085de9.png)

**Comment by wwmm** [comment]
>Also I just tested on an Ubuntu live usb and I can confirm that the "Monitor of [Device]" devices are not made by OBS, they immediately exist on the system.

Ok. OBS desktop stream records from the sound card device monitors as I was expecting. These monitors are created by the sound server. @ShayBox while a better solution is not available you should be able to workaround this problem using the per application switch in PulseEffects window. Go to the application section under the microphone tab
![Screenshot from 2021-04-12 11-33-02](https://user-images.githubusercontent.com/16225172/114411901-fe557800-9b82-11eb-9b61-924f60645ef1.png)
identify OBS desktop stream and disable its switch like in the example above.

**Comment by HritwikSinghal** [comment]
> > Also I just tested on an Ubuntu live usb and I can confirm that the "Monitor of [Device]" devices are not made by OBS, they immediately exist on the system.
> 
> Ok. OBS desktop stream records from the sound card device monitors as I was expecting. These monitors are created by the sound server. @ShayBox while a better solution is not available you should be able to workaround this problem using the per application switch in PulseEffects window. Go to the application section under the microphone tab
> ![Screenshot from 2021-04-12 11-33-02](https://user-images.githubusercontent.com/16225172/114411901-fe557800-9b82-11eb-9b61-924f60645ef1.png)
> identify OBS desktop stream and disable its switch like in the example above.
> 
> But have in mind that you will have to do that every time you open OBS if you keep `Process All Inputs` turned on...

This worked like charm! Thank you!

**Comment by ShayBox** [comment]
This worked, Turning off both switches, clicking add to blacklist on both, going into the settings and making sure `OBS` was on input and output blacklist, then restarting obs and pulseeffects

**Comment by Wuestengecko** [comment]
Hello, I'm currently facing this same issue and was hoping that my proposed solution would ultimately make everyone happy. (tl;dr at the end.)

My setup
========

- Pipewire v0.3.42 as audio server, with PW-Pulse and PW-ALSA
- EasyEffects v6.1.5 installed from repos
- Various applications using libpulse for audio input/output
- Arch Linux btw

Most of audio stuff just works™, however I'm facing the exact issue discussed here when it comes to OBS and Pavucontrol (which are the only programs I use that want to capture more than a single microphone).

Analysis of the issue
================

A picture says more than a thousand words, so, to establish a common baseline, here's a screenshot of [Helvum](https://gitlab.freedesktop.org/ryuukyu/helvum) when EasyEffects is completely turned off (as in, not running at all). A simple pipeline with two audio producers on the speakers (mpv and a qemu VM), one consumer on the microphone (the same VM) and OBS recording it all.

![Helvum with EE not running](https://user-images.githubusercontent.com/1579756/146804480-16fa2ac8-4f34-428d-8865-34f7f5db9f78.png)

Things become much busier once EE is started (with only one filter each on input/output in this case). "Process all inputs/outputs" were both off during start, and I have toggled the two other programs using the GUI. At this point, OBS has not been touched at all and it's still happily capturing the hardware devices, as expected.

![Helvum with EE switched on manually](https://user-images.githubusercontent.com/1579756/146805251-03608799-5747-4071-abf8-c50a7a4b1320.png)

OBS' own settings allow me to switch its capture to the virtual devices created by EE – and at this point, everything would be working. Great! Right? Well, not so fast. Obviously switching to the virtual sink in OBS will only capture applications feeding into that, meaning only ones I have explicitly toggled on in EE. That's quite bothersome, so let's look at the blacklist approach instead.

Without having touched OBS settings yet (it's still capturing from "default" for both desktop audio and mic), I will:

1. Add OBS to EE's blacklist (both input and output, for good measure, even though it doesn't do any playback)
2. Toggle on the "process all inputs/outputs" switches
3. Restart EE and OBS for good measure

![Helvum with OBS blacklisted](https://user-images.githubusercontent.com/1579756/146809538-feacf414-2949-4c35-b4cd-b52f0e704544.png)

Now this ... is completely unexpected. For some reason, OBS – still set up to capture from hardware – is now capturing the EE virtual sink monitor for both desktop audio *and* microphone! And even weirder: If I redo the same steps with Pavucontrol instead of OBS, all of its capture inputs end up going to the physical microphone! (You were talking about [hard-coding rules for OBS](https://github.com/wwmm/easyeffects/issues/807#issuecomment-817406330) earlier in this discussion; I presume this is why these two behave differently here.)

Conclusion
=========

The way I see it, this is clearly a bug in EasyEffects. I would expect "add to blacklist" in EE to mean "don't touch this program's streams, even though all other streams will be moved to the virtual devices". In actuality it seems to mean "move this program to the physical input / output device, instead of the virtual ones like the other programs" (at least in combination with the global switches being on) – which doesn't make sense in any scenario I can imagine.

Proposed solution (aka: tl;dr)
=======================

In the context of "Process all inputs / outputs" being on, please change the meaning of blacklisting a program from "always move this program to the physical device" to "don't touch this program at all".

**Comment by wwmm** [comment]
>The way I see it, this is clearly a bug in EasyEffects

Not necessarily. There are many situations where the audio server is the one moving streams to our devices even if we did not ask it to do that. It dependeds on whether or not you have ever set our virtual devices as default. If you did not enable "Process All Outputs" or inputs easyeffects is not the one moving your streams.

**Comment by wwmm** [comment]
>please change the meaning of blacklisting a program from "always move this program to the physical device" to "don't touch this program at all".

That happens only when the blacklist checkbutton is activated for the first time. When the application is opened after that we do not try to move it to our virtual devices. So there is no need to move it to the default device. 

But like I said before PipeWire media session manager is probably the one deciding to automatically move streams to our virtual device. With `Process All  ***` disabled the only things we try to link are our own filters.

**Comment by wwmm** [comment]
But back to the original topic the real issue is audio application delelopers not using the tags that audio servers already provide. Each application can set unique id strings to each of their streams. We could easily blocklist a given application stream if their developers were providing unique id. But as the servers do not require that they do not do it. And as a result we do not have a reliable way to do per stream blocklisting. As things are right now the only realizable blocklist has to target the application name.

**Comment by wwmm** [comment]
>In actuality it seems to mean "move this program to the physical input / output device, 

I am looking at our code again and there may have be a regression at some point at how this blocklist is handled. I will do some tests later.

**Comment by Wuestengecko** [comment]
> Pipewire v0.3.42 as audio server

I forgot to mention Wireplumber v0.4.5, everything from the repos, just in case it's relevant. But I was able to reproduce the same behavior with `pipewire-media-session` as well.

> It depends on whether or not you have ever set our virtual devices as default.

To eliminate that factor, I have removed everything pipewire and wireplumber from `~/.config` and `~/.local/state` and fully rebooted, but (almost) the exact same thing happens: Without EasyEffects running, OBS connects fine, but with EE, OBS gets moved – curiously now to the microphone (instead of the virtual sink monitor), so apparently there was something else messing around as well. But again, both of OBS' inputs land there, even though one should go to the hardware sink monitor.

I also made another intersting observation, which you can see in this short screencap of Helvum when EasyEffects starts while OBS is already running: https://drive.google.com/file/d/1AwVJ9n8vq6kELFR11CXxoMYpf5dK2fZ7/view?usp=sharing

You can clearly see a delay of ~0.5s between the filters / virtual devices appearing and the streams being moved. This suggests to me that it is indeed not the audio server itself moving things, because it would move instantly as soon as the new device appears. And this is also the only instance I could find of such a delay; everything else (i.e. programs starting into an existing filter chain) happens instantly, or at least faster than Helvum can render.

Now it is very much possible that there is actually still something else messing around, but at this point I don't know what else I could try to further isolate the cause.

> We could easily blocklist a given application stream if their developers were providing unique id.

Of course that would be the ideal solution, but my issue is about the application-level blocklist not working properly. :)

> we may have changed the blocklist code in the past in a way where we make sure the stream does not stay in our devices if it was blocklisted

I can manually move programs around with e.g. pavucontrol, and they stay where I put them, no matter whether they're listed or not.

**Comment by wwmm** [comment]
@Wuestengecko are you able to test our master branch? I have made some updates that should fix the issue about blocklisted applications being forced to go to the default device. The side effect is that if for some reason a blocklisted application is moved to our virtual device it will stay there and effects will be applied to it. But that is easier to handle that we breaking custom routing.

**Comment by Wuestengecko** [comment]
It seems that with a855254d everything is now working properly! :tada:

Note that I did have to reset wireplumber's state once more and reboot, otherwise it would still occasionally move streams (I couldn't really find a pattern here though). Anyway, using OBS' settings to select audio devices works fine now, and Pavucontrol is able to monitor everything properly now as well. Thanks so much!

**Comment by wwmm** [comment]
Despite some of the bullshit I had to read today #1430 #1431 I took a look at our blocklist code and our current master branch will use the `application.id` tag instead of the `application.name` if the streams sets it. So now it is up to OBS developers to make good use of the id tag. People curious about these tags can take a look at https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/src/pipewire/keys.h.

After today's bullshit I decided there is zero chance a workaround for OBS is done. But if they start to use `application.id` and our blocklist code does not work when trying to block specific streams let me know. Obviously.

**Comment by yavko** [comment]
> Despite some of the bullshit I had to read today #1430 #1431 I took a look at our blocklist code and our current master branch will use the `application.id` tag instead of the `application.name` if the streams sets it. So now it is up to OBS developers to make good use of the id tag. People curious about these tags can take a look at [gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/src/pipewire/keys.h](https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/src/pipewire/keys.h).
> 
> After today's bullshit I decided there is zero chance a workaround for OBS is done. But if they start to use `application.id` and our blocklist code does not work when trying to block specific streams let me know. Obviously.

I can't believe people can be like this to people who spend hours creating free and open source work. Truly disgusting.

**Comment by Digitalone1** [comment]
As I understood, EasyEffects needs some OBS streams to be "tagged" so that they can be "excluded" and not handled by EasyEffects.

@wwmm was also talking about a per-stream blocklist. Ideally OBS should define some names and give one of them to different types of stream, so EasyEffects can exclude a specific type of stream rather than everything from OBS.

I see OBS is defining the following properties:

https://github.com/obsproject/obs-studio/blob/master/libobs/audio-monitoring/pulse/pulseaudio-wrapper.c#L110-L112

```c
pa_proplist_sets(p, PA_PROP_APPLICATION_NAME, "OBS");
pa_proplist_sets(p, PA_PROP_APPLICATION_ICON_NAME, "obs");
pa_proplist_sets(p, PA_PROP_MEDIA_ROLE, "production");
```

So the application name is `OBS`. If we want to exclude OBS entirely, we could do it just adding `OBS` to the blocklist (or the new Excluded Apps List). Unless Pipewire is doing something odd and it's not setting `OBS` but something else. Unfortunately OBS is using Pulseaudio, not Pipewire.

So if Pipewire is not setting it, the issue is related to the Pulseaudio compatibility layer developed by Pipewire. Otherwise, if Pipewire is setting it correctly, just adding `OBS` should exclude OBS once and for all and the `application.id` is not even needed.

If we need a way to do per-stream exclusion, than the `application.id` could be useful if OBS sets specific id for specific type of stream. But the problem here is another. Does Pulseaudio have the `application.id` property to tag a specific type of stream?

If yes, a merge request can be made to change the sign of the following function
```c
pa_stream *pulseaudio_stream_new(const char *name, 
                                 const pa_sample_spec *ss,
			         const pa_channel_map *map)
```
so that it takes as parameter also another string `type_name` and set it in `pulseaudio_properties()` accordingly. Then every call to `pulseaudio_stream_new` should be changed to specify also the type. This way we can exclude specific stream when clicking on the `Exclude` checkbutton (former `Add to Blocklist`), rather than everything from OBS.

**Comment by wwmm** [comment]
>If we need a way to do per-stream exclusion, than the application.id could be useful if OBS sets specific id for specific type of stream. But the problem here is another. Does Pulseaudio have the application.id property to tag a specific type of stream?

I do not think the fact they are using Pulseaudio will be a problem. PulseEffects sets the `application.id` tag and PAvucontrol does the same thing to its probe streams. This works on Pulseaudio. OBS just has to set it in the same way they set `PA_PROP_APPLICATION_NAME`.

**Comment by wwmm** [comment]
Pulseaudio has this tag https://gitlab.freedesktop.org/pulseaudio/pulseaudio/-/blob/master/src/pulse/proplist.h#L147. PipeWire code seems to have reused this file. So it is totally doable by OBS. All they have to do is setting one `id string` to the stream that has to record from the microphone and another id to the one that records from the soundcard monitor. This way EasyEffects users will be able to blocklist only one of these streams instead of all OBS streams.

**Comment by wwmm** [comment]
Anyone curious about the `application.id` usage on the audio application side can take a look at the output of `pactl list source-outputs` while Pavucontrol is running and something is playing audio. In the output you will see something like 
```
Properties:
		client.api = "pipewire-pulse"
		pulse.server.type = "unix"
		application.name = "PulseAudio Volume Control"
		application.id = "org.PulseAudio.pavucontrol"
		application.icon_name = "audio-card"
```

Pavucontrol sets `application.id = "org.PulseAudio.pavucontrol"` to its probe streams. OBS just has to do the same but with different values for each of its 2 streams. This way we will know which one we have to block.

At this moment they use the `media name tag` for this. But as I have said before audio and video players use this tag for something else. And on top of that the media tag can be translated. It is a horrible tag when trying to do a blocklist.

**Comment by Digitalone1** [comment]
Sure. I was thinking `obsproject.obs-studio.input-monitor` and `obsproject.obs-studio.output-monitor`.

After completing the new app info ui and fixing the locale issue, I'll try to download obs git and make a merge request. Shouldn't be so much difficult to catch all function calling `pulseaudio_stream_new` and set the proper id. Or at least I hope so.

**Comment by Digitalone1** [comment]
So I took a look at obs code.

The pulse properties are set while creating a new pulse stream with `pa_stream_new_with_proplist`

- https://github.com/obsproject/obs-studio/blob/bf660b1d8dc1905527bb5919b1034c7b43c55dac/libobs/audio-monitoring/pulse/pulseaudio-wrapper.c#L300
- https://github.com/obsproject/obs-studio/blob/bf660b1d8dc1905527bb5919b1034c7b43c55dac/plugins/linux-pulseaudio/pulse-wrapper.c#L264

This call is made inside 2 functions 

- `pulseaudio_stream_new` in pulseaudio-output.c
https://github.com/Digitalone1/obs-studio/blob/bf660b1d8dc1905527bb5919b1034c7b43c55dac/libobs/audio-monitoring/pulse/pulseaudio-output.c#L478

- `pulse_stream_new` in pulse-input.c
https://github.com/obsproject/obs-studio/blob/bf660b1d8dc1905527bb5919b1034c7b43c55dac/plugins/linux-pulseaudio/pulse-input.c#L362

Which are respectively inside `audio_monitor_init` and `pulse_start_recording`:

- https://github.com/obsproject/obs-studio/blob/bf660b1d8dc1905527bb5919b1034c7b43c55dac/libobs/audio-monitoring/pulse/pulseaudio-output.c#L392
- https://github.com/obsproject/obs-studio/blob/bf660b1d8dc1905527bb5919b1034c7b43c55dac/plugins/linux-pulseaudio/pulse-input.c#L329

Someone may say, the first in pulseaudio-output is for output device and the second in pulse-input is for input device, but I'm not really sure. I don't even know if at the moment of invoking `pa_stream_new_with_proplist` there is meaningful data to distinguish between an input and an output stream, maybe the functions containing them are made to work on both types. 

So maybe the stream type is known outside and the only way to set a proper id is to pass the type by parameter. Anyway it seems too complicated for me, so I need an hint or the issue has to be submitted to obs devs.

**Comment by wwmm** [comment]
I did not have time to look at their code. But the call using `obs_source_get_name(monitor->source)` may be the one that creates the stream responsible for recording the sound card monitors. Or at least the `monitor` object suggests that is what is being done.

What I tried to find but I did not have time to is where OBS is setting the `media name`. In Pavucontrol we can see that give different names to the media tag.

**Comment by wwmm** [comment]
When OBS is running the output of `pactl list source-outputs` shows for one stream `media.name = "Mic/Aux"` and for the other `media.name = "Desktop Audio"`. So the place where they are doing that may be the place where to set the id.

**Comment by wwmm** [comment]
>So the place where they are doing that may be the place where to set the id

Or maybe not. They seem to be setting the media name at a later time through a call to `obs_source_set_name`. I do not know if the same approach can be done for other tags like the `app id`.

**Comment by Digitalone1** [comment]
Searched through the whole project, there's no place using the `Desktop Audio` and `Mix/Aux` strings.

The only properties they are setting "directly" are `PA_PROP_APPLICATION_NAME`, `PA_PROP_APPLICATION_ICON_NAME` and `PA_PROP_MEDIA_ROLE`.

Anyway take a look at this function:
- https://github.com/obsproject/obs-studio/blob/bf660b1d8dc1905527bb5919b1034c7b43c55dac/plugins/linux-pulseaudio/pulse-input.c#L457

pulse_properties. On the `input` bool parameter they are calling one of the following: 

- https://github.com/obsproject/obs-studio/blob/bf660b1d8dc1905527bb5919b1034c7b43c55dac/plugins/linux-pulseaudio/pulse-wrapper.c#L166
- https://github.com/obsproject/obs-studio/blob/bf660b1d8dc1905527bb5919b1034c7b43c55dac/plugins/linux-pulseaudio/pulse-wrapper.c#L188

I don't understand very much, but does it seem they are somehow copying the properties from the existing source/sink?

**Comment by Digitalone1** [comment]
When I have time I will open a ticked on OBS github. Maybe they can suggest a method to do this.

**Comment by wwmm** [comment]
>but does it seem they are somehow copying the properties from the existing source/sink?

In this case `pulse_properties` is definitely getting properties from a source/sink. But it is not clear to me yet if the result is being added to the stream properties. Maybe it is.

**Comment by Digitalone1** [comment]
Unfortunately reporting a bug on their Github is a little bit tricky. I need a log which is mandatory, but I don't have any and don't want to install the application to generate a new one. If someone could provide one here while the issue is occurring, it would be very appreciated and I will complete the ticket on OBS issue section. Thanks.

**Comment by wwmm** [comment]
The person responsible for #1430 has apologized #1541 and provided an idea that could be used in a workaround that is decent enough to be used. It is already in our master branch so it would be good to know from the OBS users if the problem is fixed.

Honestly it is not as good as making OBS developers setting a batter tag because users won't be able to customize this through our blocklist. The fix is hardcoded in our sources. But it is life...

**Comment by sephid86** [comment]
> > Despite some of the bullshit I had to read today #1430 #1431 I took a look at our blocklist code and our current master branch will use the `application.id` tag instead of the `application.name` if the streams sets it. So now it is up to OBS developers to make good use of the id tag. People curious about these tags can take a look at [gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/src/pipewire/keys.h](https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/src/pipewire/keys.h).
> > After today's bullshit I decided there is zero chance a workaround for OBS is done. But if they start to use `application.id` and our blocklist code does not work when trying to block specific streams let me know. Obviously.
> 
> I can't believe people can be like this to people who spend hours creating free and open source work. Truly disgusting.

부끄럽다.
I felt ashamed.

좋은 모습을 보이지 못한것에 대해 너에게도 미안하다.
I'm also sorry for not showing you a good side.

나는 니가 나와는 다르게 좋은 어른이 되기를 바란다.
I want you to become a good adult unlike me.

**Comment by wwmm** [comment]
EasyEffects `6.2.6` has been released today with the workarounds I talked about in the posts above. This issue can be closed.
